R/list-transpose.R

```diff
@@ -0,0 +1,267 @@
+#'
+#' - `vec_ptype(list_transpose(x)[[i]]) == vec_ptype_common(!!!x, .ptype = ptype)`
+#' - `vec_size(list_transpose(x)[[i]]) == vec_size(x)`
+#'
+#' If `NULL` elements are present in `x`, then an error is thrown unless `null`
+#' is provided, in which case any `NULL` elements are treated as size 1 for the
+#' common size computation.
+#'
+#' @export
+#' @examples
+#' # Input:
+#' # - List size 3
+#' # - Element size 2
+#' # Output:
+#' # - List size 2
+#' # - Element size 3
+#' list_transpose(list(1:2, 3:4, 5:6))
+#'
+#' # With data frames
+#' x <- data_frame(a = 1:2, b = letters[1:2])
+#' y <- data_frame(a = 3:4, b = letters[3:4])
+#' list_transpose(list(x, y))
+#'
+#' # Size 1 elements are recycled to the common size before transposing
+#' list_transpose(list(1, 2:4))
+#'
+#' # With all size 1 elements, you can use `size` if you want to force a known
+#' # common size other than size 1
+#' list_transpose(list(1, 2), size = 3)
+#'
+#' # With size 0 elements, the invariants are a bit tricky!
+#' # This must return a size 0 list, but then you lose expected
+#' # type (integer) and size (2) information about the elements.
+#' # Losing that information makes it difficult to reverse the
+#' # transposition.
+#' #
+#' # Input:
+#' # - List size 2
+#' # - Element size 0
+#' # Output:
+#' # - List size 0
+#' # - Element size 2
+#' x <- list(integer(), integer())
+#' out <- list_transpose(x)
+#' out
+#'
+#' # Note how transposing a second time doesn't recover the original list
+#' list_transpose(out)
+#'
+#' # To work around this, provide the lost `size` and `ptype` manually
+#' list_transpose(out, size = vec_size(x), ptype = vec_ptype_common(!!!x))
+#'
+#' # If you'd like to pad with a missing value rather than recycling or
+#' # erroring, you might do something like this, which left-pads
+#' x <- list(1, 2:5, 6:7)
+#' try(list_transpose(x))
+#'
+#' sizes <- list_sizes(x)
+#' size <- max(sizes)
+#' index <- which(sizes != size)
+#' x[index] <- lapply(
+#'   index,
+#'   function(i) vec_c(rep(NA, times = size - sizes[[i]]), x[[i]])
+#' )
+#' list_transpose(x)
+#'
+#' # `NULL` values aren't allowed in `list_transpose()`
+#' x <- list(1:3, NULL, 5:7, NULL)
+#' try(list_transpose(x))
+#'
+#' # Use `null` to replace `NULL` values before transposing
+#' list_transpose(x, null = NA)
+#' list_transpose(x, null = 0L)
+list_transpose <- function(
+  x,
+  ...,
+  null = NULL,
+  size = NULL,
+  ptype = NULL,
+  x_arg = caller_arg(x),
+  error_call = current_env()
+) {
+  check_dots_empty0(...)
+
+  obj_check_list(x, arg = x_arg, call = error_call)
+
+  if (is.object(x)) {
+    # The list input type should not affect the transposition process in any
+    # way. In particular, supplying a list subclass that doesn't have a
+    # `vec_cast.subclass.list` method shouldn't prevent the insertion of
+    # `list(null)` before the transposition. The fact that we must insert
+    # `list(null)` should be considered an internal detail.
+    x <- unclass(x)
+  }
```

Comment:
I have tests for this, but basically I want this to work

```r
x <- structure(list(1, NULL), class = c("my_list", "list"))
list_transpose(x, null = 0)
```

But it doesn't due to `null` handling because we have to do this

``` r
vec_assign(x, vec_detect_missing(x), list(0))
#> Error in `vec_assign()`:
#> ! Can't convert <list> to <my_list>.
```

which doesn't work unless you bypass the class

We use `obj_check_list()` to check that it's a list, and after that we don't care about the class type.